xlb/operator/collision/smagorinsky_les_bgk.py

```diff
@@ -0,0 +1,143 @@
+import jax.numpy as jnp
+from jax import jit
+import warp as wp
+from typing import Any
+import numpy as np
+
+from xlb.velocity_set import VelocitySet
+from xlb.compute_backend import ComputeBackend
+from xlb.operator.collision.collision import Collision
+from xlb.operator import Operator
+from functools import partial
+
+
+class SmagorinskyLESBGK(Collision):
+    """
+    BGK collision operator for LBM with Smagorinsky LES model.
+    """
+
+    def __init__(
+        self,
+        velocity_set: VelocitySet = None,
+        precision_policy=None,
+        compute_backend=None,
+        smagorinsky_coef: float = 0.17,
+    ):
+        self.smagorinsky_coef = smagorinsky_coef
+        super().__init__(velocity_set, precision_policy, compute_backend)
+
+    def _construct_warp(self):
+        # Set local constants TODO: This is a hack and should be fixed with warp update
+        _w = self.velocity_set.w
+        _c = self.velocity_set.c
+        _smagorinsky_coef = wp.constant(self.compute_dtype(self.smagorinsky_coef))
+        _f_vec = wp.vec(self.velocity_set.q, dtype=self.compute_dtype)
+
+        # Construct the functional
+        @wp.func
+        def functional(
+            f: Any,
+            feq: Any,
+            rho: Any,
+            u: Any,
+            omega: Any,
+        ):
+            # Compute the non-equilibrium distribution
+            fneq = f - feq
+
+            # Sailfish implementation
+            #{
+            #  float tmp, strain;
+
+            #  strain = 0.0f;
+
+            #  // Off-diagonal components, count twice for symmetry reasons.
+            #  %for a in range(0, dim):
+            #    %for b in range(a + 1, dim):
+            #       tmp = ${cex(sym.ex_flux(grid, 'd0', a, b, config), pointers=True)} -
+            #           ${cex(sym.ex_eq_flux(grid, a, b))};
+            #       strain += 2.0f * tmp * tmp;
+            #    %endfor
+            #  %endfor
+
+            #  // Diagonal components.
+            #  %for a in range(0, dim):
+            #    tmp = ${cex(sym.ex_flux(grid, 'd0', a, a, config), pointers=True)} -
+            #        ${cex(sym.ex_eq_flux(grid, a, a))};
+            #    strain += tmp * tmp;
+            #  %endfor
+
+            #  tau0 += 0.5f * (sqrtf(tau0 * tau0 + 36.0f * ${cex(smagorinsky_const**2)} * sqrtf(strain)) - tau0);
+            #}
+
+            # Compute strain
+            strain = wp.float32(0.0)
```

Comment:
This could be an incorrect implementation. It seems very stable and I dont notice too many differences with KBC though.